nav2_experimental/nav2_rl/nav2_turtlebot3_rl/turtlebot3_env_core.py

```diff
@@ -0,0 +1,312 @@
+# Copyright (c) 2019 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from time import sleep
+from threading import Thread
+
+import rclpy
+from rclpy.node import Node
+from rclpy.duration import Duration
+from rclpy.qos import qos_profile_sensor_data
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.parameter import Parameter
+
+import numpy as np
+from math import pi, atan2, sin, cos
+import math
+import os
+import random
+import pandas
+import parameters
+
+from geometry_msgs.msg import Twist, Pose
+from sensor_msgs.msg import LaserScan
+from std_srvs.srv import Empty
+from std_msgs.msg import String
+from gazebo_msgs.srv import GetEntityState, SetEntityState
+import copy
+
+
+class TurtlebotEnv(object):
+    def __init__(self, TurtlebotEnv=None):
+        self.TurtlebotEnv = TurtlebotEnv
+        self.node_ = rclpy.create_node('turtlebot3_env_core')
+        self.executor = SingleThreadedExecutor()
+        self.executor.add_node(self.node_)
+        self.node_.declare_parameter('use_sim_time', True)
+        self.node_.set_parameters([Parameter('use_sim_time', Parameter.Type.BOOL, True)])
+        self.act = 0
+        self.done = False
+        self.actions = self.get_actions()
+
+        self.collision = False
+        self.collision_tol = 0.125
+        self.laser_scan_range = [0] * 360
+        self.states_input = [3.5] * 8
+        self.zero_div_tol = 0.01
+        self.range_min = 0.0
+        self.states = []
+
+        self.current_pose = Pose()
+        self.goal_pose = Pose()
+
+        self.pub_cmd_vel = self.node_.create_publisher(Twist, 'cmd_vel', 1)
+        self.sub_scan = self.node_.create_subscription(LaserScan, 'scan', self.scan_callback,
+                                                       qos_profile_sensor_data)
+
+        self.reset_simulation = self.node_.create_client(Empty, 'reset_simulation')
+        self.reset_world = self.node_.create_client(Empty, 'reset_world')
+        self.unpause_proxy = self.node_.create_client(Empty, 'unpause_physics')
```

Comment:
I dont see you pausing the physics between runs. This is necessary to ensure the training is running at the correct rate with the correct propogation of physics between them.

Ei unpause -> act on environment -> let run for 1/frequency time -> pause -> return result. That way the simulation is stopping between updates and not continuing to creep.